[CLO-40] - Make updateFromJson doc save optional

sebialex opened the pull request.

sebialex committed:
[CLO-40] - created 1.3.6-gene42 patient-data-default-impl
- added loadFromJson() mehtod to PhenoTipsPatient

buske reviewed (approved):
Add a TODO to refactor to separate the responsibilities when we migrate to pt-core-1.4, and then it looks good to me

sebialex committed:
[CLO-40] - overloaded updateFromJSON in PhenoTipsPatient
- the new method now takes a boolean which if set, only triggers a save if the document ends up dirty after the controllers run their course

danielpgross commented on components/patient-data/impl/src/main/java/org/phenotips/data/internal/PhenoTipsPatient.java:
```diff
@@ -279,12 +292,17 @@ public void updateFromJSON(JSONObject json)
         return toJSON().toString(2);
     }
+
+    private static boolean saveIfDirty(XWikiDocument document, boolean saveOnlyIfDirty)
```
This method name is confusing to me because it implies that the method actually saves the document. How about `shouldSaveDocument` instead?

sebialex replied:
I'll think about it some more, but I didn't want to call it that because the first boolean 'save' serves that purpose, and the method deals with dirty flag. If I called it `shouldSaveDocument`, then I might as well have brought in the first boolean as well, and that would have defeated the purpose.

sebialex replied:
My concern is that the dirty flag is always true due to some other mechanism, even if nothing changed.

danielpgross commented on components/patient-data/impl/src/main/java/org/phenotips/data/internal/PhenoTipsPatient.java:
```diff
@@ -267,7 +267,20 @@ private void updateFromJSON(JSONObject json, boolean save)
+     */
+    public void updateFromJSON(JSONObject json, boolean saveOnlyIfDirty)
+    {
+        this.updateFromJSON(json, true, true);
```
Shouldn't the third argument be `saveOnlyIfDirty`?

sebialex replied:
Yes

danielpgross reviewed (changes requested):
Looks good, aside from these minor changes.

One of us should run through some common cases with the debugger to ensure that the document's `isDirty` flag is being activated on any change within a controller.

sebialex committed:
[CLO-40] - fixed overloaded updateFromJSON

sebialex committed:
[CLO-40] - renamed saveIfDirty to shouldSaveIfDirty
